src/ert/ensemble_evaluator/monitor_async.py

```diff
@@ -0,0 +1,135 @@
+from websockets.client import connect
+
+from ert.ensemble_evaluator import identifiers
+from ert.ensemble_evaluator._wait_for_evaluator import wait_for_evaluator
+from ert.serialization import evaluator_marshaller, evaluator_unmarshaller
+
+if TYPE_CHECKING:
+    from ert.ensemble_evaluator.evaluator_connection_info import EvaluatorConnectionInfo
+
+
+logger = logging.getLogger(__name__)
+
+
+class MonitorAsync:
+    def __init__(self, ee_con_info: "EvaluatorConnectionInfo") -> None:
+        self._ee_con_info = ee_con_info
+        self._id = str(uuid.uuid1()).split("-", maxsplit=1)[0]
+        self._event_queue: asyncio.Queue[CloudEvent] = asyncio.Queue()
+        self._connection: Optional[WebSocketClientProtocol] = None
+        self._receiver_task: Optional[asyncio.Task[None]] = None
+        self._connected: asyncio.Event = asyncio.Event()
+
+    async def __aenter__(self) -> "MonitorAsync":
+        self._receiver_task = asyncio.create_task(self._receiver())
+        await self._connected.wait()
+        return self
+
+    async def __aexit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        if self._connection:
+            await self._connection.close()
+        if self._receiver_task:
+            if not self._receiver_task.done():
+                self._receiver_task.cancel()
+            # we are done and not interested in errors when cancelling
+            await asyncio.gather(
+                self._receiver_task,
+                return_exceptions=True,
+            )
+
+    def get_base_uri(self) -> str:
+        return self._ee_con_info.url
+
+    async def signal_cancel(self) -> None:
+        if not self._connection:
+            return
+        logger.debug(f"monitor-{self._id} asking server to cancel...")
+
+        out_cloudevent = CloudEvent(
+            {
+                "type": identifiers.EVTYPE_EE_USER_CANCEL,
+                "source": f"/ert/monitor/{self._id}",
+                "id": str(uuid.uuid1()),
+            }
+        )
+        await self._connection.send(
+            to_json(out_cloudevent, data_marshaller=evaluator_marshaller)
+        )
+        logger.debug(f"monitor-{self._id} asked server to cancel")
+
+    async def signal_done(self) -> None:
+        if not self._connection:
+            return
+        logger.debug(f"monitor-{self._id} informing server monitor is done...")
+
+        out_cloudevent = CloudEvent(
+            {
+                "type": identifiers.EVTYPE_EE_USER_DONE,
+                "source": f"/ert/monitor/{self._id}",
+                "id": str(uuid.uuid1()),
+            }
+        )
+        await self._connection.send(
+            to_json(out_cloudevent, data_marshaller=evaluator_marshaller)
+        )
+        logger.debug(f"monitor-{self._id} informed server monitor is done")
+
+    async def track(self) -> AsyncGenerator[CloudEvent, None]:
+        while True:
+            event = await self._event_queue.get()
```

Comment:
Can this potentially hang if there are no more events in event queue?
In AsyncEE, you had this wrapped in an `asyncio.wait_for` with a timeout to make sure it would not hang. Can we implement something similar here?

Also, can we change `while True` -> `while self._is_running`, or would this not bring any benefit?

Reply:
You might be correct, actually maybe the `while True | self._running` might not be even required (need to test it). The point is to wait for another event - blocked fashion, which might take some time until it arrives. So I wouldn't do `wait_for`. It's a context manager that kills the tasks once exiting.

Reply:
Ok, checked and it is required.